src/async_stub.rs

```diff
@@ -0,0 +1,93 @@
+use std::collections::VecDeque;
+use std::io;
+use std::time::{Duration, Instant};
+
+use futures::{
+    channel::mpsc::{self as channel, UnboundedSender},
+    prelude::*,
+    StreamExt,
+};
+
+use crate::{MspPacket, MspPacketParseError, MspParser};
+
+pub struct AsyncMspStub {
+    tx: UnboundedSender<Request>,
+}
+
+struct Request {
+    message: MspPacket,
+    backchannel: UnboundedSender<Result<MspPacket, MspPacketParseError>>,
+}
+
+struct Pending {
+    cmd: u16,
+    backchannel: UnboundedSender<Result<MspPacket, MspPacketParseError>>,
+    ts: Instant,
+}
+
+impl AsyncMspStub {
+    pub fn new<R: AsyncRead + Unpin + Send, W: AsyncWrite + Unpin + Send>(
+        mut reader: R,
+        mut writer: W,
+        timeout: Duration,
+    ) -> io::Result<(Self, impl Future<Output = impl Send> + Send)> {
+        let (tx, mut rx) = channel::unbounded();
+        let f = async move {
+            let mut parser = MspParser::new();
+            let mut buf = [0u8; 1];
+            let mut pending: VecDeque<Pending> = VecDeque::with_capacity(16);
+
+            loop {
+                futures::select! {
+                    // maybe a new response?
+                    _ = reader.read_exact(&mut buf[..]).fuse() =>
+                           match parser.parse(buf[0]){
+                               Ok(Some(response))=>{
+                                while let Some(p) = pending.front_mut(){
+                                    if p.ts.elapsed()> timeout {
+                                        p.backchannel.send(Err(MspPacketParseError::TimedOut)).await.unwrap();  // TODO convert to an err
+                                        pending.pop_front();
+                                    }else if p.cmd == response.cmd{
+                                        p.backchannel.send(Ok(response)).await.unwrap(); // TODO convert to an err
+                                        pending.pop_front();
+                                        break;
+                                    }
+                               }
+                               }
+                               _=>{
+                                   // TODO what now?
```

Comment:
We would need to bubble any parsing errors up to the caller, maybe with separate err channel?

Reply:
I think it is better to use a Result channel directly, if we bubble up. However, if we assume that 

+ some message was dropped (CRC error, someone tickles the serialport header, whatever),
+ we don't know how many messages where dropped

Then we can impossibly yield the IO error precisely to that pending request which it belongs to. The question is, what do we do then? We could send an IO Error to all pending requests. However, if we do so we might cancle request which where not dropped to which we will receive a response in the near future. So really I figured, the only sane thing to do here is to ignore the error, eventually the timeout will kick in and our client will know that _something_ went wrong. I know, far from ideal.

Reply:
Yes, I faced similar problem and just send the error messages, so the user of the api can print it or do something with it... but because the async nature there is no good way of know which error correspond to which message.
Which makes me wonder if the msp protocol needs to to support some sort of id to be able to work asynchronously. 
I guess msp never meant to be async protocol and it must work in serial request response fashion.
And i should stop abusing it to do something it can't do.